### PR TITLE
CLI output selection in assembler mode

### LINKED ISSUE #3870: [CLI] Do not ignore the output selection in assembler mode
I am using the `--assemble` mode for transforming assembly-syntax into binary representation through a tool.

This works good except for the fact that the output of the compiler is not really machine-readable as it just dumps everything to stdout and stderr.

It would be nice if `solc` would care about the `--bin` flag that is already present if run in assemble mode (`--assemble`).

This way, one would not have to use a regex over the data in stdout in order to retrieve the binary representation.

Current behaviour:
```
➜  /tmp solc --assemble --bin temp_2.asm

======= temp_2.asm (EVM) =======

Pretty printed source:
{
    0x40
    0x50
    add
    pop
}

Binary representation:
604060500150

Text representation:
    /* "temp_2.asm":3:7   */
  0x40
    /* "temp_2.asm":9:13   */
  0x50
    /* "temp_2.asm":15:18   */
  add
    /* "temp_2.asm":20:23   */
  pop
```

Desired behaviour:
```
➜  /tmp solc --assemble --bin temp_2.asm
604060500150
```

### PR DESCRIPTION
**Pull request by cameel**
Resolves #3870.
~Depends on #12071. In draft mode until that PR is merged.~ Merged.

This does not add any new outputs, just makes the ones that are already available respond to output selection flags so e.g. `--ir` is not available. We can 

Also, to make this non-breaking, all outputs are selected if no flags are specified, which is different from how the compiler mode behaves. If this goes through, I can create a separate PR to `breaking` branch to change that.

### CODE REVIEW

**Comment by cameel** [comment]
#12071 merged so this is now reviewable.

**Comment by cameel** [comment]
Note: this is now based on the CI fix (#12106).

File: solc/CommandLineInterface.cpp [hunk 1]
```diff
@@ -1013,34 +1013,54 @@ bool CommandLineInterface::assemble(yul::AssemblyStack::Language _language, yul:
...
 
 		if (_language != yul::AssemblyStack::Language::Ewasm && _targetMachine == yul::AssemblyStack::Machine::Ewasm)
 		{
 			stack.translate(yul::AssemblyStack::Language::Ewasm);
 			stack.optimize();
 
-			sout() << endl << "==========================" << endl;
-			sout() << endl << "Translated source:" << endl;
-			sout() << stack.print() << endl;
+			// TODO: This isn't ewasm but it's only present when we're doing Yul->EWASM translation.
```
Comment (leonardoalt) [inline solc/CommandLineInterface.cpp]: I guess this is not going to be done in this PR? Maybe create an issue so it doesn't get forgotten?

File: solc/CommandLineParser.cpp [hunk 2]
```diff
@@ -497,6 +504,16 @@ bool CommandLineParser::parseOutputSelection()
 	for (auto&& [optionName, outputComponent]: CompilerOutputs::componentMap())
 		m_options.compiler.outputs.*outputComponent = (m_args.count(optionName) > 0);
 
+	if (m_options.input.mode == InputMode::Assembler && m_options.compiler.outputs == CompilerOutputs{})
+	{
+		// In assembly mode keep the default outputs enabled for backwards-compatibility.
+		// TODO: Remove this (must be done in a breaking release).
```
Comment (leonardoalt) [inline solc/CommandLineParser.cpp:487]: Maybe also create issue?

**Review by leonardoalt — DISMISSED** [review]

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (cameel) [reply solc/CommandLineInterface.cpp]: Will do.

File: solc/CommandLineParser.cpp (hunk 2 above)
Comment (cameel) [reply solc/CommandLineParser.cpp:487]: Will do.

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (cameel) [reply solc/CommandLineInterface.cpp:1029]: Actually, I just realized that ewasm is still experimental. So maybe it's not breaking after all?

File: solc/CommandLineParser.cpp (hunk 2 above)
Comment (cameel) [reply solc/CommandLineParser.cpp:510]: Actually, I just realized that ewasm is still experimental. So maybe it's not breaking after all?

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (cameel) [reply solc/CommandLineInterface.cpp]: Actually, I just realized that ewasm is still experimental. So maybe it's not breaking after all?

File: solc/CommandLineParser.cpp (hunk 2 above)
Comment (cameel) [reply solc/CommandLineParser.cpp:487]: #12230.

File: solc/CommandLineInterface.cpp (hunk 1 above)
Comment (cameel) [reply solc/CommandLineInterface.cpp]: Got a confirmation on the chat. It's not breaking.

In that case I'm adding an extra commit that adds an output called `--ewasm-ir` (I could not come up with a better name).

**Review by leonardoalt — DISMISSED** [review]

**Comment by leonardoalt** [comment]
wait, do the tests need update?

**Comment by cameel** [comment]
I did update them but looks like I missed one change needed after a rebase. Apparently an extra `stop` instruction gets added after `sstore` due to #11974. There are other similar changes in that PR so it seems legit.

Should be fixed now.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**